### PR TITLE
2.0 Changes

### PR DESCRIPTION
**Pull request by jamesjamail**
Lots of changes...

-Debounced multiple watch notifications for the same seller + item combo within a 15 minute window
-Added "block seller" feature for individual items/watches or account wide
-Added "snooze" feature for individual items/watches or account wide
-Added emoji functionality through out
-Reworked command system
-Watch Notifications and !watches results now attempt to fetch the wiki item image if possible
-Watch Notifications scrape wiki for auction data if possible
-Various formatting, copy, and color changes
 
alterTables.sql is the handwritten file I created by copying the sql statements out of pgAdmin4 as I was making the changes.  I've also included a dump of schema only.

### CODE REVIEW

File: README.md
Comment (jamesjamail) [inline README.md:18]: <3

File: client.js
Comment (jamesjamail) [inline client.js:56]: I hope this will be a pretty seamless transition.  If a user enters an old command, bot lets them know and shows the new commands.

File: client.js
Comment (jamesjamail) [inline client.js:69]: I found myself still using colons a lot, so I wanted the command to still execute but inform the user.

File: client.js
Comment (jamesjamail) [inline client.js:165]: Using emojis in other commands too now, it's a much nice experience.  I'm wondering if the 24 hour period on these event listeners will be expensive cpu-wise...

File: client.js
Comment (jamesjamail) [inline client.js:391]: This is where I'm checking for debounce limit, snoozes, or blocks.

File: client.test.js
Comment (jamesjamail) [inline client.test.js:23]: Don't remember doing this...never got around to the integrated testing.

File: logParser.js
Comment (jamesjamail) [inline logParser.js:5]: Figured I could increase this without impacting performance too much.

File: package.json
Comment (jamesjamail) [inline package.json:34]: Using moment to list watch expirations as how much time is left rather than a fixed date.  Moment is dying soon RIP.

File: wikiHandler.js
Comment (jamesjamail) [inline wikiHandler.js:88]: I got lazy here and didn't fully explore your code.  formatted_items is doing a lot that just gets thrown away.  TODO.

File: messages.js
Comment (rm-you) [inline messages.js:34]: Not super important, but trying to remember if this will auto-link the channel or if there's something we need to do to make that a channel link...

File: schema_files/alterTables.sql
Comment (rm-you) [inline schema_files/alterTables.sql:1]: So, this file should at least be numbered like "update_001" or something... and maybe the TQB_Dump.sql should be like "base" or "update_000" or something so it's clear what order to run things in? Assuming that is a dump *before* these updates...

File: utils.js
Comment (rm-you) [inline utils.js:50]: Wait, so this is not dynamic (not sure if we CAN do stuff that's dynamic when it's just for a message) but you've just chosen to optimize for mobile and leave it at that? I guess I'm fine with this, just curious since you put the note for the desktop value here.

File: wikiHandler.js
Comment (rm-you) [inline wikiHandler.js:62]: Just actually noticed that the Auctioneer is only in the footer now -- I guess this is fine, though I kinda liked having it more visible I guess? And it's not like we're short on the space in that line, since it is a whole line by itself, plenty of space. You really prefer not to put the name there?

File: wikiHandler.js
Comment (rm-you) [inline wikiHandler.js:63]: Interesting that Timestamp still only shows the date on Desktop, but shows the date+time on mobile! Cool feature, should handle the complaints about mobile formatting not showing time easily I guess. :P

File: wikiHandler.js
Comment (rm-you) [reply wikiHandler.js:88]: Uhh yeah, I'm trying to figure out what is changing but my brain hurts a little. Don't see anything missing though in the test server so maybe it's ok? Not sure what would be missing...

File: wikiHandler.js
Comment (rm-you) [inline wikiHandler.js:196]: So wait, where do these images go? I don't actually see them anywhere...

File: utils.js
Comment (rm-you) [inline utils.js:77]: Err, please don't do things this way -- we literally have items.json for this purpose. Just look up the key/value to get the wiki URL.

File: utils.js
Comment (rm-you) [inline utils.js:63]: I believe if you look up the item in the items.json  in all-caps by key, you can take the "wiki URL" (value) and replace underscores with spaces, and you have the correct capitalization?

File: wikiHandler.js
Comment (rm-you) [inline wikiHandler.js:24]: Sorry for the bikeshedding, but... I liked the old colors :(
RIP. What was the motivation for changing these?

**Review by rm-you — CHANGES_REQUESTED** [review]
How much I want to use https://sequelize.org/ is off the charts >_<
That can happen later, doesn't have to be in this change, but jeebus. ORM for the win.

**Comment by rm-you** [comment]
Oh, you still haven't mocked-out the redis connection stuff, so it's trying to contact redis in the tests (which will obviously fail). Need to fix that so the testing is meaningful again.

File: wikiHandler.js
Comment (jamesjamail) [reply wikiHandler.js:24]: I needed a third color for system related messages like !list and !help and didn't like any of the hues that went with the original server colors.  At the same time, I get several users who claim their watch notifications aren't coming through because they saw a tunnel stream post on the other server for their item.  I wanted to see if making the colors brighter would make it more obvious which server they are looking at.  Also just think these look slicker, but that's just my opinion.  Can put it up as a poll and see what users want.

File: wikiHandler.js
Comment (jamesjamail) [reply wikiHandler.js:196]: They go in the response to !watches and watch notifications.  If the item name doesn't match a url, it isn't displayed.  Figured I could improve that later.

File: utils.js
Comment (jamesjamail) [reply utils.js:63]: I use this for most everything, like prices and server names too.

File: utils.js
Comment (jamesjamail) [reply utils.js:77]: Good call, I'll take a look.

File: client.js
Comment (jamesjamail) [reply client.js:544]: Ah, good catch.  My bad.

I made a renewWatch function which I'll put here.  I'm surprised I missed that here.

File: client.js
Comment (jamesjamail) [reply client.js:366]: =)

File: client.js
Comment (jamesjamail) [reply client.js:87]: Nice ><

File: db.js
Comment (jamesjamail) [reply db.js:350]: I was trying really hard to get this all in one query with a bunch of joins, but wasn't able to get it working reliably.  Looking at it now I could just omit the else statements since the if statements always return- that would keep it from nesting horizontally.

File: messages.js
Comment (jamesjamail) [reply messages.js:34]: It does not auto link the channel, I thought it would =/

File: schema_files/alterTables.sql
Comment (jamesjamail) [reply schema_files/alterTables.sql:1]: That's a good call.  The TQB_Dump.sql reflects the same schema as this, but is just the auto generated sql dump in case I missed something in this 'hand-written' alterTables.sql file.  I was considering having Mtil dump his db so that I could test that these changes in alterTables.sql work, but that juice might not be worth the squeeze.  Agree we should use sequelize or something in the future.

File: utils.js
Comment (jamesjamail) [reply utils.js:50]: Man, sorry- I had a lot of old code I didn't realize that I should have deleted before pushing.  This was an experiment.  I was trying to write a function that would be like a 'float right' with padding spaces before the word, but because it's not a mono-spaced font it doesn't work well.  Can ignore- will delete.

File: wikiHandler.js
Comment (jamesjamail) [reply wikiHandler.js:62]: I did a lot of tossing and turning over this, but I do think it's best this way.  In my opinion, having just the auction type in the headline is cleaner and makes it easier to quickly look a stream and pick out which is message is.  The timestamp at the bottom is necessary for mobile users who can't inspect the time for each message, so adding the auctioneer here isn't wasted space either.  My thoughts were that the only time a user cares who the auctioneer is is when they actually want to buy/sell the item, which is rare compared to wanting to know the auction type, items, prices, and other info.  I also think having the auctioneer by the timestamp makes sense together..."Rabitz on Project 1999 Blue | 10:27am" is all related info in my eyes.  Can create a poll for it too, I hear your point that it's easier to read in the title- I could be wrong that users want a less cluttered experience.

File: wikiHandler.js
Comment (jamesjamail) [reply wikiHandler.js:63]: The most annoying part of this is that on mobile it uses a bullet before the time and on desktop it uses a | .  But yeah, I think it'll be good.  It will also reliably format to user's local timezone which I'm not sure we would be able to do within our own code.

File: wikiHandler.js
Comment (rm-you) [reply wikiHandler.js:196]: Oh, ok they show up when you do a WATCH. I was looking for them in the auction stream. 👍

File: utils.js
Comment (rm-you) [reply utils.js:63]: Ah, interesting... Ok. Might be fine, but IMO would be better then to switch the specific locations where we know it's item names to use the method I'm talking about, rather than using this function. But, maybe for display purposes it's just "good enough".

File: wikiHandler.js
Comment (rm-you) [reply wikiHandler.js:24]: Ah, I thought maybe the green change was a reference to Teal, since it was so light. 🤣 
I wouldn't waste too much time on it, it's really just bikeshedding. Was just curious.
I think the "other server confusion" thing happens when people use search and Discord doesn't automatically add the `in:#p99_green_tunnel_stream` filter for some reason. I've talked to a couple of people who were confused because of this. Not sure if we can fix this easily...

File: utils.js
Comment (rm-you) [reply utils.js:77]: This reminds me, we should fix the item name match for that item someone mentioned that was wrong -- was it Executioner's Axe? Can fix simply by changing the key in items.json, which will make the lookup match properly and link to the wiki correctly regardless of the difference.

File: client.js
Comment (jamesjamail) [reply client.js:56]: ><

File: logParser.js
Comment (jamesjamail) [reply logParser.js:5]: Fixed.

File: package.json
Comment (jamesjamail) [reply package.json:34]: I've been using dayjs and it's very light weight and similar, so that's an option for the future.  Marking resolved for now.

File: wikiHandler.js
Comment (jamesjamail) [reply wikiHandler.js:88]: Nothing is missing- it's actually the opposite: formatItemData also returns some values that are not being used.  Minor performance issue if we want to fix in the future.  I was thinking of including a  "price check" command at a later date, so we can just address this then.

File: utils.js
Comment (jamesjamail) [reply utils.js:77]: Yeah- the user was searching for "Executioners Axe" without an apostrophe.  Should I add these keys then?
```
{
"EXECUTIONER'S AXE":"/Executioner%27s_Axe" //we already have this one
"EXECUTIONERS AXE":"/Executioner%27s_Axe",
"AN EXECUTIONER'S AXE":"/Executioner%27s_Axe" ,
"AN EXECUTIONERS AXE":"/Executioner%27s_Axe",
}
```

File: wikiHandler.js
Comment (jamesjamail) [reply wikiHandler.js]: I made Green's color a bit darker

File: messages.js
Comment (rm-you) [reply messages.js:34]: There IS a way to make it link the channel, but it requires a little bit of code. Can worry about it later.

File: utils.js
Comment (rm-you) [reply utils.js]: Perfect, in aliases like you did, yes!

File: client.js
Comment (rm-you) [inline client.js:7]: Why is this a second line of the same import? Is it a style thing? Any performance repercussions for doing it this way? I don't know how it works in JS, in Python this WOULD be totally normal/fine to do, but in some other languages a double import is not good. No clue.

File: client.js
Comment (rm-you) [inline client.js:199]: This feature is cool, but isn't this the wrong place for this? It looks like this took the place of "un-refresh", which is fine because that seems like a kinda weird/nonsensical option, but this should really be on an individual watch notification, right? Not the watch show/list command?

File: client.js
Comment (rm-you) [inline client.js:202]: Also, isn't this totally the wrong message for "unblock"?

File: client.js
Comment (rm-you) [inline client.js:322]: Does the output of this command allow you to un-do the blocks in any way? I'm not clear how this would work, as it looks like right now you're building a single big message?
Should it respond with one message per block, and have an emoji to press to undo that block?

File: client.js
Comment (rm-you) [inline client.js:369]: lolwat

Anyway, missing punctuation if you happen to make another pass.

File: client.js
Comment (rm-you) [inline client.js:447]: That "alt" isn't valid anymore as you've chosen to use that emoji for "silence seller", so having it listed here confused me for a moment

File: client.js
Comment (rm-you) [inline client.js:481]: Wrong message, and doesn't do anything? This should be "now watching this auction again", and it should START a watch. You've just left "endWatch" commented...

File: client.js
Comment (rm-you) [inline client.js:486]: Wrong message.

File: client.js
Comment (rm-you) [inline client.js:500]: Wait, what? I'm finding it hard to process what this if/else was for... It looks like it's a TON of repeated code? What is it actually handling, the case where the user isn't able to be found in cache?
```
    if (bot.users.cache.get(user.toString()) === undefined) {
        bot.guilds.cache.get(GUILD).members.fetch(user.toString())
            .then((res)=>{
                res.send(msg)
                # More stuff happens
            }
    } else {
        bot.users.cache.get(user.toString()).send(msg)
        # Same stuff as above happens
    }
```

If that's it, this could be handled MUCH more cleanly -- just make the if/else grab the user, and once you have it, do all of this stuff to the result... Something like:
```
    if (bot.users.cache.get(user.toString()) === undefined) {
        bot.guilds.cache.get(GUILD).members.fetch(user.toString())
            .then((res)=>{
                actuallySendMessageFunction(res, msg);
            }
    } else {
        actuallySendMessageFunction(bot.users.cache.get(user.toString()), msg);
    }

function actuallySendMessageFunction(user_obj, message) {
    # All of that other stuff that was duplicated before
}
```

File: client.js
Comment (rm-you) [reply client.js:548]: Looks like you didn't actually resolve this. But, per above, this would be a duplicate if you combined the code-paths.

File: client.js
Comment (rm-you) [inline client.js:554]: Same as above, but should go away if combined.

File: db.js
Comment (rm-you) [inline db.js:182]: I really hope this properly handles SQL Escaping, I will cry at you if this is a SQL Injection path, since all of the item names are user input.

**Review by rm-you — CHANGES_REQUESTED** [review]
Some work required...

So, for the SQL files you're adding -- one will go from blank-DB to schema 001, and the other will go from "original schema" to 001? Meaning we just don't care about the original schema, and no one will ever need to use it or understand it again? :D
Just checking.

File: client.js
Comment (jamesjamail) [reply client.js:7]: I didn't notice, good catch.  Shouldn't error in JS but no reason it shouldn't be one import.

File: client.js
Comment (jamesjamail) [reply client.js:199]: Good catch, copy/pasta gone wrong.  Removing the "remove" event listener for  '🔕' which doesn't belong here anyway..

File: client.js
Comment (jamesjamail) [reply client.js:202]: Yup.

File: client.js
Comment (jamesjamail) [reply client.js:322]: That's a great idea, let's add it to Issues as a feature to do.

File: client.js
Comment (jamesjamail) [reply client.js:369]: Good catch.

File: client.js
Comment (jamesjamail) [reply client.js:486]: Excellent catch.

File: db.js
Comment (jamesjamail) [reply db.js:182]: connection.query sanitizes the inputs in the second argument (the array) =)

**Comment by jamesjamail** [comment]
> Some work required...
> 
> So, for the SQL files you're adding -- one will go from blank-DB to schema 001, and the other will go from "original schema" to 001? Meaning we just don't care about the original schema, and no one will ever need to use it or understand it again? :D
> Just checking.

Exactly.  I will have Mtil try to use db_update_001 to preserve current data, but if that doesn't work for whatever reason, we at least have an failsafe to make the upgrade even if we lose user data in the process.

File: client.js
Comment (jamesjamail) [reply client.js:554]: 👍

File: client.js
Comment (jamesjamail) [reply client.js:500]: You're right, you're right =)

File: client.js
Comment (jamesjamail) [reply client.js:447]: Oops, my bad.

File: client.js
Comment (rm-you) [inline client.js:464]: Message here is now wrong.

Comment (jamesjamail) [reply client.js:464]: Good catch.

File: wikiHandler.js
Comment (jamesjamail) [inline wikiHandler.js:132]: This was the offender.  I assumed redis functions would already be promisified but turns out they aren't.

File: wikiHandler.js
Comment (jamesjamail) [inline wikiHandler.js:190]: Changed your await placement a bit.

**Review by rm-you — APPROVED** [review]
Alright, this looks... Good? I will admit to a *bit* of review fatigue, so it's possible I missed something... If everything is working for you on the test server though, I guess we're ready to give this a shot.

I'd recommend doing a DB restore from your 000 base (the one posted in another PR that we never merged) and then testing your 000->001 sql script one more time just to catch anything obvious that snuck in, before we risk having to walk mtil through fixing some random failure. Hopefully is fine though. 👍

**Comment by rm-you** [comment]
You may need to disable the branch protection requirement of passing CI tests, before you can merge? Unless you have permissions to override that and just click the Squash+Merge button (I do not).

**Comment by jamesjamail** [comment]
> Alright, this looks... Good? I will admit to a _bit_ of review fatigue, so it's possible I missed something... If everything is working for you on the test server though, I guess we're ready to give this a shot.
> 
> I'd recommend doing a DB restore from your 000 base (the one posted in another PR that we never merged) and then testing your 000->001 sql script one more time just to catch anything obvious that snuck in, before we risk having to walk mtil through fixing some random failure. Hopefully is fine though. 👍

Understandable, yes- there are lots of things that could be improved but I think this is fine for the moment.  Last night I was running the bot in the VS Code's terminal, then VS Code crashed but apparently the shell process kept running in the background.  I was debugging for hours why messages were being doubled up and console.log weren't making any sense before I checked task manager. Big oof.  Once I realized that, I retested everything and it's all working very nicely.  The bot is really in good shape now, I think users will really appreciate the work we've done.

Thanks for making time to help with this review.  I know I was initially wanting to rush these changes out, but I think it's good you got me to slow down a little bit because there's things I definitely would have missed.

Will merge and try to perform the db changes locally before giving mtil the green light.

**Pull request merged**
